---
title: "Data QC and QA Part 2"
output: html_notebook
---

## Part 2 - Sample level QC

```{r}
#add new column for capture

pData(BS.obj)$capture <- as.numeric(gsub("^[A-Z](\\d+)-.*$", "\\1", pData(BS.obj)$Sample_Names))

metadata$capture <- pData(BS.obj)$capture

# View the updated dataframe
head(pData(BS.obj))
head(metadata)

save(BS.obj, metadata, file = file.path(outdir, "BS.obj.rda"))
```


### DensityPlot of methylated reads

```{r}
#Density plot
densityPlot(bsseq::getMeth(BS.obj, type = 'raw')) #any failed samples should be obviously different from the rest

#creating groups with suspect samples
groups = which(pData(BS.obj)$BioMood.ID[] %in% c('BioMood24', 'BioMood35', 'BioMood50', 'BioMood39', 'BioMood52'))

groups = pData(BS.obj)$BioMood.ID[which(pData(BS.obj)$BioMood.ID[] %in% c('BioMood24', 'BioMood35', 'BioMood50', 'BioMood39', 'BioMood52'))]

#Coloring Density plot by capture pools
densityPlot(bsseq::getMeth(BS.obj, type = 'raw'), sampGroups = pData(BS.obj)$capture, pal = brewer.pal(8, "Accent"))

#Coloring Density plot by sample IDs
densityPlot(bsseq::getMeth(BS.obj, type = 'raw'), sampGroups = pData(BS.obj)$BioMood.ID, pal = colorRampPalette(brewer.pal(8, "Dark2"))(52))

pdf(file=paste0(plotdir,sep="/","densityplot_bycapture.pdf"), width=10)
```

### Sex check
```{r}
#get sex chromosomes
chrX = chrSelectBSseq(BS.obj, seqnames = c("X"))
chrY = chrSelectBSseq(BS.obj, seqnames = c("Y"))

#get the methylated reads on x and y
x.meth = bsseq::getMeth(chrX, type='raw')
y.meth = bsseq::getMeth(chrY, type='raw')

#average methylation over x and y
xMed = matrixStats::colMedians(x.meth)
yMed = matrixStats::colMedians(y.meth)

#Predict sex based on the difference
predictedSex <- ifelse(yMed - xMed < 0.25, "F", "M")

#create a data frame for data viz
df <- as.data.frame(cbind('sample' = colnames(xMed), xMed, yMed))
#df$sex = pData(BS.obj)$sex #here we use the sex assignments in the subject metadata as 
df$sex <- rep('F', 52)
df$predictedSex <- predictedSex
```


```{r fig.width=6, fig.height=6}
ggscatter(df, x = 'xMed', y = 'yMed', color = "sex", shape="predictedSex",
      palette = c("#00AFBB", "#ff69b4"), size=3, alpha=0.5)+
  ggsci::scale_colour_npg()+
  labs(title = 'Sex prediction plot', x = "Median intensity chrX", y = "Median intensity chrY") + 
   theme_bw(base_rect_size = 2, base_size = 20)
```


# Save .rda object and session
Saving the annotated BS.obj into output directory which is ready for downstream analysis.

```{r warning =FALSE, echo=FALSE, message=FALSE}

save(BS.obj, metadata, file = file.path(outdir, "BS.obj.rda"))
sessionInfo()

```

#Session info
```{r}
sessionInfo()
```



