---
title: "Exploratory Analysis"
output: html_notebook
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

These scripts assume methyldackel calls have been processed and QC'd into filtered bsseq object for data analysis. Here we get a general overview of major sources of variation in the methylation data and we will estimate cell composition for each sample. 

Requirements
-------------
1. Processed data set stored as .rda object
2. Annotation file for TWIST methylation panel

# Setup

## load libraries
```{r, message=FALSE, warning=FALSE}
library(bsseq)
library(tidyverse)
library(plyranges)
library(ggpubr)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
library(EpiDISH)
library(annotatr)
library(org.Hs.eg.db)
```

## directories
```{r}
datadir <- '/mnt/scratch/BioMood'
metadir <-"~/file_workspace/Collaborative Projects/HPC_CEPI/ANALYSIS/BioMood_Epi/Subject_info" 
plotdir <- "/mnt/scratch/BioMood/plots"
resdir <- "/mnt/scratch/BioMood/results"
refdir <- "/mnt/scratch/BioMood/references"
```

# Load methylation data and annotations
```{r}
load(file.path(datadir,'BS.obj.rda'))
     
annotation.file <- list.files(file.path(refdir), recursive=TRUE, pattern="*hg19_annotated_collapsed_uniq_final.bed", full=TRUE)

#we create an annotation object called m2g (stands for map to genome)
m2g <- read.table(
annotation.file,
sep= "\t",
col.names = c("chr", "start", "end", "annotation"),
colClasses = c("character", "integer", "integer", "character"))

```

# Remove outlier samples - BioMood39, BioMood50, BioMood24, BioMood35

```{r}

id.mis <- grep("A5-BioMood39rpt|B6-BioMood50|H6-BioMood24|F6-BioMood35", metadata$Sample_Name) #get the locations of samples with suspect data

metadata_quality <- metadata[-id.mis,] #remove them from metadata table
BS.obj_quality <- BS.obj[,-id.mis]    #remove them from sequencing data

save(BS.obj_quality, metadata_quality, file = file.path(outdir, "BS.obj.rda"))

```


# Gene promoters
```{r}
genes = build_annotations(genome = 'hg19', annotations =c('hg19_basicgenes','hg19_genes_intergenic'))

promoters <- subset(genes, type=="hg19_genes_promoters")

```

# CpG Islands
```{r}
cgislands = build_annotations(genome = 'hg19', annotations =c("hg19_cpg_islands","hg19_cpg_shelves", "hg19_cpg_shores"))
```

# Enhancer regions
```{r}
# Build the annotations (a single GRanges object)
enhancers = build_annotations(genome = 'hg19', annotations = "hg19_enhancers_fantom")
```


```{r}
library(plyranges)

seqlevelsStyle(enhancers) <- "NCBI"
seqlevelsStyle(promoters) <- "NCBI"
seqlevelsStyle(cgislands) <- "NCBI"

#with removed samples
bs.enhancers <- subsetByOverlaps(BS.obj_quality, enhancers, minoverlap = 1)

bs.promoter <-subsetByOverlaps(BS.obj_quality, promoters)

bs.cgi <- subsetByOverlaps(BS.obj_quality, cgislands)

```


# Exploratory analysis - PCA Gene Promoters

```{r}
#extract methylation ratios
Beta <-  bsseq::getMeth(bs.promoter, type='raw') %>%
  as.matrix() %>% round(2)

features <- as.data.frame(bs.promoter@rowRanges)

rownames(Beta) = paste(features$seqnames,features$start,sep=":")
```



```{r}
#Principal components 
pc <- prcomp(t(Beta), center = TRUE, scale = F,retx=TRUE)

#Extract the loadings or the rotated data and the standard deviation
loadings=pc$x
sd = pc$sdev
```


```{r, fig.width=10, fig.height=6}
# Add row number to metadata
metadata_with_rownum <- metadata %>%
  dplyr::mutate(RowNumber = row_number()) %>%
  dplyr::select(RowNumber, BioMood.ID, BioMood.group, everything())

metadata_with_rownum$RowNumber <- as.character(metadata_with_rownum$RowNumber)

metadata_with_rownum$capture <- as.factor(metadata_with_rownum$capture)

metadata_with_rownum$BioMood.group <- as.factor(metadata_with_rownum$BioMood.group)

pca.raw = loadings %>%
  as_tibble(rownames = 'RowNumber') %>%
  dplyr::select('RowNumber', PC1, PC2) %>%
  dplyr::left_join(metadata_with_rownum) %>%
  ggscatter(x = 'PC1', y = 'PC2', color = 'capture',
      shape = 'BioMood.group', alpha=0.5, label='BioMood.ID') +
  ggsci::scale_colour_npg()+
  labs(title = 'PCA Gene promoters', x = "Dim1", y = "Dim2") + 
   theme_bw(base_rect_size = 2, base_size = 20)
pca.raw

pdf(file=paste0(plotdir,sep="/","PCA_methylation_prom_bycapture.pdf"), width=10)
pca.raw
dev.off()

```

# Exploratory analysis - PCA Gene Enhancers

```{r}
#extract methylation ratios
Beta <-  bsseq::getMeth(bs.enhancers, type='raw') %>%
  as.matrix() %>% round(2)

features <- as.data.frame(bs.enhancers@rowRanges)

rownames(Beta) = paste(features$seqnames,features$start,sep=":")
```


```{r}
#Principal components 
pc <- prcomp(t(Beta), center = TRUE, scale = F,retx=TRUE)

#Extract the loadings or the rotated data and the standard deviation
loadings=pc$x

```


```{r, fig.width=10, fig.height=6}
metadata_with_rownum$BioMood.group <- as.factor(metadata_with_rownum$BioMood.group)

pca.raw = loadings %>%
  as_tibble(rownames = 'RowNumber') %>%
  dplyr::select('RowNumber', PC1, PC2) %>%
  dplyr::left_join(metadata_with_rownum) %>%
  ggscatter(x = 'PC1', y = 'PC2', color = "capture",
     shape = 'BioMood.group', alpha=0.5 , label='BioMood.ID') +
  ggsci::scale_colour_npg()+
  labs(title = 'PCA Gene Enhancers', x = "Dim1", y = "Dim2") + 
   theme_bw(base_rect_size = 2, base_size = 20)
pca.raw

pdf(file=paste0(plotdir,sep="/","PCA_methylation_enh_bycapture.pdf"), width=10)
pca.raw
dev.off()

```

# Exploratory analysis - PCA CGIslands

```{r}
#extract methylation ratios
Beta <-  bsseq::getMeth(bs.cgi, type='raw') %>%
  as.matrix() %>% round(2)

features <- as.data.frame(bs.cgi@rowRanges)

rownames(Beta) = paste(features$seqnames,features$start,sep=":")
```


```{r}
#Principal components 
pc <- prcomp(t(Beta), center = TRUE, scale = F,retx=TRUE)

#Extract the loadings or the rotated data and the standard deviation
loadings=pc$x

```


```{r, fig.width=10, fig.height=6}

pca.raw = loadings %>%
  as_tibble(rownames = 'RowNumber') %>%
  dplyr::select('RowNumber', PC1, PC2) %>%
  dplyr::left_join(metadata_with_rownum) %>%
  ggscatter(x = 'PC1', y = 'PC2', color = "capture",
      shape = 'BioMood.group', alpha=0.5, label='BioMood.ID') +
  ggsci::scale_colour_npg()+
  labs(title = 'PCA CpG Islands', x = "Dim1", y = "Dim2") + 
   theme_bw(base_rect_size = 2, base_size = 20)
pca.raw

pdf(file=paste0(plotdir,sep="/","PCA_methylation_cgi_bycapture.pdf"), width=10)
pca.raw
dev.off()

```

#Principal components analysis of batch effects
```{r}
library(minfi)
library(RColorBrewer)
library(limma)
library(ggplot2)
library(WGCNA)
library(sva)

# scree plot and variance explained
par(mfrow = c(1,2))
plot(pc,type="lines", main = 'scree plot')
barplot(sd[1:10]/sd[1], main = 'variance explained') 

pdf(file=paste0(plotdir,sep="/","Scree_plot_variance_explained.pdf"), width=10)

dev.off()
```

#Correlation heatmap
```{r}
#Matrix of covariates
datTraits=data.frame(Diet = factor(metadata_quality$BioMood.group),
                    BMI =factor(metadata_quality$BMIprepreg),
                    Age = factor(metadata_quality$Age))
```

```{r}
#Variables to specify
nSamples = ncol(M)

datTraits_numeric <- datTraits %>%
  mutate(across(where(is.factor), ~ as.numeric(as.character(.))))

# Pearson Correlation of the first 5 PC with the supplied covariates
moduleTraitCor = cor(loadings[,c(1:5)], datTraits_numeric, use='complete.obs', method = "pearson")
moduleTraitPvalue = corPvalueStudent(moduleTraitCor, nSamples)

#Spearman Correlation of the first 5 PC with the supplied covariates
moduleTraitCor = cor(loadings[,c(1:5)], datTraits_numeric, use='complete.obs', method = "spearman")
moduleTraitPvalue = corPvalueStudent(moduleTraitCor, nSamples)

# Will display correlations and their p-values
textMatrix = paste(signif(moduleTraitCor, 2), "\n(",
                   signif(moduleTraitPvalue, 1), ")", sep = "");
                   dim(textMatrix) = dim(moduleTraitCor)

par(mar = c(6, 8.5, 3, 3));

# Display the correlation values within a heatmap plot
labeledHeatmap(Matrix = t(moduleTraitCor),
               xLabels = colnames(loadings)[1:5],
               yLabels = names(datTraits),
               colorLabels = FALSE,
               colors = blueWhiteRed(8),
               textMatrix = t(textMatrix),
               setStdMargins = FALSE,
               cex.text = 0.6,
               zlim = c(-1,1),
               main = paste("PCA-trait relationships"))

pdf(file=paste0(plotdir,sep="/","correlation_heatmap.pdf"), width=10)

dev.off()
```

#MDS analysis
```{r}
#parameters to set
pal = colorRampPalette(brewer.pal(8,"Dark2"))(8)
pal = colorRampPalette(brewer.pal(16, name = "Oranges"))
dims = list(c(1,2),c(1,3),c(2,3),c(3,4))
par(mfrow=c(2,2), oma=c(0,0,1,0))

#function to run
sapply(1:length(dims), function(i){
  plotMDS(M, 
          top=5000, 
          gene.selection="common", 
          dim.plot=dims[[i]],
          pch = ifelse(datTraits$Diet==1, 16,17), 
          col=pal[factor(metadata_quality$Age)])
})

pdf(file=paste0(plotdir,sep="/","MDS_analysis.pdf"), width=10)

dev.off()

```
```{r}
```


#Session Info
```{r}
sessionInfo()
```